### PR TITLE
boards: add initial support for the PINE64 PineTime smartwatch

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

I just got the PineTime devkit. It was a breeze to port, from opening the device to hello world took <15minutes. Thanks @aabadie @haukepetersen et al for the nice abstractions there!

For more information about the board, please look here:
https://wiki.pine64.org/index.php/PineTime

*edit*
This currently only adds the basic board configuration and flashing etc. No drivers are used, yet.

### Testing procedure

AFAIK they just started shipping the devkits, so I'm probably the only maintainer who has one.
I'll provide output and test results.

For now, maybe concentrate on the integration.

### CODE REVIEW

**Comment by kaspar030** [comment]
examples/hello-world output:

```
[kaspar@ng riot/examples/hello-world (add_pinetime_support)]$ BOARD=pinetime make term
/home/kaspar/src/riot/dist/tools/jlink/jlink.sh term_rtt
### Starting RTT terminal ###
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2019-10-23 13:52:20,988 # Host name for TCP connection is missing, defaulting to "localhost"
2019-10-23 13:52:20,988 # Connect to localhost:19021
Welcome to pyterm!
Type '/exit' to exit.
2019-10-23 13:52:21,993 # SEGGER J-Link V6.44e - Real time terminal output
2019-10-23 13:52:21,993 # SEGGER J-Link EDU V10.1, SN=260107551
2019-10-23 13:52:21,994 # Process: JLinkExe
2019-10-23 13:52:21,994 # main(): This is RIOT! (Version: 2020.01-devel-284-gdb33e-add_pinetime_support)
2019-10-23 13:52:21,994 # Hello World!
2019-10-23 13:52:21,995 # You are running RIOT on a(n) pinetime board.
2019-10-23 13:52:21,995 # This board features a(n) nrf52 MCU.
```

Other than that, I tested xtimer_msg, which works fine.

File: dist/tools/codespell/check.sh [hunk 1]
```diff
@@ -38,7 +38,7 @@ ${CODESPELL_CMD} --version &> /dev/null || {
 CODESPELL_OPTS="-q 2"  # Disable "WARNING: Binary file"
 CODESPELL_OPTS+=" --check-hidden"
 # Disable false positives "nd  => and, 2nd", "WAN => WANT", "od => of"
-CODESPELL_OPTS+=" --ignore-words-list=ND,nd,WAN,od"
+CODESPELL_OPTS+=" --ignore-words-list ND,nd,wan,od,dout"
```
Comment (cgundogan) [inline dist/tools/codespell/check.sh]: Huh? Was this supposed to end up in this PR?

Comment (kaspar030) [reply dist/tools/codespell/check.sh]: no, sorry, thanks!

**Comment by kaspar030** [comment]
- force-pushed removal of the unrelated commit

**Comment by bergzand** [comment]
Awesome!

Skimming through the display controller datasheet, it looks like it is compatible with the display driver from #9948

**Comment by kaspar030** [comment]
> Skimming through the display controller datasheet, it looks like it is compatible with the display driver from #9948

Yeah! I've already configured the settings. 1 hour of RIOT hello-world drained the battery, need to recharge before I can try...

**Comment by bergzand** [comment]
> Yeah! I've already configured the settings. 1 hour of RIOT hello-world drained the battery, need to recharge before I can try...

Can we have some proof when you have the display working? :smile:

**Comment by patchedsoul** [comment]
Will this be landing in the January release by chance?

**Comment by kaspar030** [comment]
> @kaspar030 @bergzand Will this be landing in the January release by chance?

I'll put this to high-priority. With the ili9341 merged, this should be possible. ping @fjmolinas, this would be awesome to have!

**Comment by bergzand** [comment]
> I'll put this to high-priority. With the ili9341 merged, this should be possible. ping @fjmolinas, this would be awesome to have!

I should receive my PineTime in the next few days, I can test this PR as soon as I have it.

File: boards/pinetime/Makefile.include [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# for this board, we are using Segger's RTT as default terminal interface
+USEMODULE += stdio_rtt
+TERMPROG = $(RIOTTOOLS)/jlink/jlink.sh
+TERMFLAGS = term_rtt
```
Comment (bergzand) [inline boards/pinetime/Makefile.include]:
```suggestion
 TERMFLAGS = term-rtt
```

File: boards/pinetime/include/gpio_params.h [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+#define GPIO_PARAMS_H
+
+#include "board.h"
+#include "saul/periph.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/* move on, nothing to see here */
```
Comment (bergzand) [inline boards/pinetime/include/gpio_params.h]: Can you add the GPIO button available on the PineTime here?

Comment (bergzand) [inline boards/pinetime/include/gpio_params.h]: And I guess this should be extended with the vibrator GPIO as soon as SAUL supports vibrators. This can of course wait for a follow up :)

File: boards/pinetime/include/periph_conf.h [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+#ifndef PERIPH_CONF_H
+#define PERIPH_CONF_H
+
+#include "periph_cpu.h"
+#include "cfg_clock_32_1.h"
+#include "cfg_i2c_default.h"
+#include "cfg_rtt_default.h"
+#include "cfg_spi_default.h"
+#include "cfg_timer_default.h"
+
```
Comment (bergzand) [inline boards/pinetime/include/periph_conf.h]: Should the battery ADC line be included here?

File: boards/pinetime/Makefile.features [hunk 5]
```diff
@@ -0,0 +1,7 @@
+CPU_MODEL = nrf52832xxaa
+
+# Put defined MCU peripherals here (in alphabetical order)
+#FEATURES_PROVIDED += periph_i2c
+#FEATURES_PROVIDED += periph_uart
```
Comment (bergzand) [inline boards/pinetime/Makefile.features]: UART is not available on the board right?

File: boards/pinetime/Makefile [hunk 6]
```diff
@@ -0,0 +1,4 @@
+MODULE = board
+DIRS = $(RIOTBOARD)/common/nrf52xxxdk
```
Comment (bergzand) [inline boards/pinetime/Makefile]: Why is this pointed to the `nrf52xxxdk` and not `nrf52`? Is there some refactoring required in those directories?

File: boards/pinetime/Makefile.include [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+# use shared Makefile.include
+include $(RIOTBOARD)/common/nrf52xxxdk/Makefile.include
```
Comment (bergzand) [inline boards/pinetime/Makefile.include]: Same as above, including `nrf52xxxdk`-files while this is not part of the DK development kits is a bit confusing.

File: boards/pinetime/doc.txt [hunk 8]
```diff
@@ -0,0 +1,12 @@
+/**
+@defgroup    boards_pinetime PineTime
+@ingroup     boards
+@brief       Support for the Pinetime
+
+### General information
+
+"The PINE64 SmartWatch, dubbed "PineTime", is a product of a community effort for an open source smartwatch in collaboration with wearable RTOS and Linux app developers & communities."
+
+See https://wiki.pine64.org/index.php/PineTime#PineTime_DevKit_SWD_Probe for more information.
```
Comment (bergzand) [inline boards/pinetime/doc.txt]: Why link a specific section of that page here? Isn't it more safe to link to the top of the page (without `#PineTime_DevKit_SWD_Probe` in the url)

**Comment by kaspar030** [comment]
I've got some more code here: https://github.com/kaspar030/pinetime-riot

The branch includes a simple module for stdio on the display. It is based on a pre-merged ili9341, though.

**Comment by bergzand** [comment]
> I've got some more code here: https://github.com/kaspar030/pinetime-riot

If I remember correctly that's the branch I based my [firmware application] on :laughing: 

[firmware application]: https://github.com/bosmoment/PineTime-apps

**Comment by kaspar030** [comment]
> If I remember correctly that's the branch I based my [firmware application](https://github.com/bosmoment/PineTime-apps) on laughing

Wow, that's quite advanced! Looking forward to try!

File: boards/pinetime/include/gpio_params.h (hunk 3 above)
Comment (bergzand) [inline boards/pinetime/include/gpio_params.h]: Should the backlight GPIO pins also be included here. As an alternative, maybe it is possible to use PWM on them to have more granular brightness control.

Comment (kaspar030) [reply boards/pinetime/include/gpio_params.h]: Yes, I got all those in my pinetime branch. will update soon!

File: boards/pinetime/doc.txt (hunk 8 above)
Comment (kaspar030) [reply boards/pinetime/doc.txt]: Agreed, that was an accident.

File: boards/pinetime/Makefile.features (hunk 5 above)
Comment (kaspar030) [reply boards/pinetime/Makefile.features]: Isn't there a debug pin or two that could be repurposed? but yeah, not really.

File: boards/pinetime/Makefile.include (hunk 2 above)
Comment (kaspar030) [reply boards/pinetime/Makefile.include]: This is coppied, others use `term_rtt`:

```
[kaspar@ng riot (add_pinetime_support)]$ gg -i term.rtt
Makefile.include:# TERMFLAGS must be exported for `jlink.sh term_rtt`.
boards/hamilton/Makefile.include:TERMFLAGS = term_rtt
boards/pinetime/Makefile.include:TERMFLAGS = term_rtt
boards/ruuvitag/Makefile.include:TERMFLAGS = term_rtt
boards/thingy52/Makefile.include:TERMFLAGS = term_rtt
dist/tools/jlink/jlink.sh:  term_rtt)
```

File: boards/pinetime/include/gpio_params.h (hunk 3 above)
Comment (kaspar030) [reply boards/pinetime/include/gpio_params.h]: yup, done

Comment (kaspar030) [reply boards/pinetime/include/gpio_params.h]: yup, its there. the vibrator is just a GPIO. On my devkit it worked only when on the cradle, though.
Anyhow, the define should be there now.

File: boards/pinetime/Makefile.include (hunk 2 above)
Comment (bergzand) [reply boards/pinetime/Makefile.include]: And it was modified in #12733 to `term-rtt`.

Comment (kaspar030) [reply boards/pinetime/Makefile.include]: Thanks. fixed

File: boards/pinetime/board.c [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+    cpu_init();
+
+    gpio_init(VCC33, GPIO_OUT);
+    gpio_init(BUTTON0_ENABLE, GPIO_OUT);
+    gpio_set(BUTTON0_ENABLE);
+    gpio_init(BUTTON0, GPIO_IN);
+    gpio_init(VIBRATOR, GPIO_OUT);
+    gpio_init(LCD_BACKLIGHT_LOW, GPIO_OUT);
+    gpio_init(LCD_BACKLIGHT_MID, GPIO_OUT);
+    gpio_init(LCD_BACKLIGHT_HIGH, GPIO_OUT);
```
Comment (bergzand) [inline boards/pinetime/board.c]: Please be a good neighbor and disable the vibrator, and both `LCD_BACKLIGHT_HIGH` and `LCD_BACKLIGHT_MID` with `gpio_set(VIBRATOR)` (those pins are active low), preferably before initializing them as output if that is possible.

File: boards/pinetime/include/board.h [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+
+#define LCD_BACKLIGHT_LOW           GPIO_PIN(0, 14)
+#define LCD_BACKLIGHT_MID           GPIO_PIN(0, 22)
+#define LCD_BACKLIGHT_HIGH          GPIO_PIN(0, 23)
+
+#define BUTTON0_ENABLE              GPIO_PIN(0, 13)
+#define BUTTON0                     GPIO_PIN(0, 15)
+#define VIBRATOR                    GPIO_PIN(0, 16)
+
+#define VCC33                       GPIO_PIN(0, 24)
```
Comment (bergzand) [inline boards/pinetime/include/board.h]: Maybe also include:
```
#define POWER_PRESENCE              GPIO_PIN(0, 19)
#define CHARGING_ACTIVE             GPIO_PIN(0, 12)
#define BATTERY_ADC                 NRF52_AIN7
```
here

File: boards/pinetime/include/periph_conf.h [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+        .dev  = NRF_SPI0,
+        .sclk = GPIO_PIN(0, 2),
+        .mosi = GPIO_PIN(0, 3),
+        .miso = GPIO_PIN(0, 4),
+    }
+};
+
+#define SPI_NUMOF           ARRAY_SIZE(spi_config)
+/** @} */
+
```
Comment (bergzand) [inline boards/pinetime/include/periph_conf.h]: And I2C config here:
```C
/**                                                                             
 * @name    I2C configuration                                                   
 * @{                                                                           
 */                                                                             
static const i2c_conf_t i2c_config[] = {                                        
    {                                                                           
        .dev = NRF_TWIM1,                                                       
        .scl = GPIO_PIN(0, 7),                                                  
        .sda = GPIO_PIN(0, 6),                                                  
        .speed = I2C_SPEED_FAST                                                 
    }                                                                           
};                                                                              
#define I2C_NUMOF           (sizeof(i2c_config) / sizeof(i2c_config[0]))        
/** @} */                                                                       
```

**Comment by bergzand** [comment]
If I have time this evening I can see if I can also give you a snippet to add config for the SPI flash.

File: boards/pinetime/Makefile (hunk 6 above)
Comment (kaspar030) [reply boards/pinetime/Makefile]: I was re-using the board.c. Anyhow, fixed.

File: boards/pinetime/Makefile.include (hunk 7 above)
Comment (kaspar030) [reply boards/pinetime/Makefile.include]: fixed

File: boards/pinetime/board.c (hunk 9 above)
Comment (kaspar030) [reply boards/pinetime/board.c]: done

File: boards/pinetime/Makefile.features [hunk 12]
```diff
@@ -0,0 +1,8 @@
+CPU_MODEL = nrf52832xxaa
+
+# Put defined MCU peripherals here (in alphabetical order)
+#FEATURES_PROVIDED += periph_i2c
```
Comment (bergzand) [inline boards/pinetime/Makefile.features]: This doesn't have to be commented out anymore now that the I2C config is included.

Comment (kaspar030) [reply boards/pinetime/Makefile.features]: yup

**Comment by kaspar030** [comment]
- addressed all comments

The PR is now able to directly use the ili9341 if the backlight is enabled. E.g., `tests/drivers_ili9341` displays something with adding this line: https://gist.github.com/kaspar030/8e252913801e8bb26978093b85dff7f9/raw. I (conditionally) added this to that driver's main.c.
I'd also like to not use the test synchronization for the ili test, if compiling for the pinetime. If used, the pinetime waits on stdio_rtt for 's' indefinitely.

**Comment by bergzand** [comment]
> The PR is now able to directly use the ili9341 if the backlight is enabled. E.g., `tests/drivers_ili9341` displays something with adding this line: https://gist.github.com/kaspar030/8e252913801e8bb26978093b85dff7f9/raw. I (conditionally) added this to that driver's main.c.
> I'd also like to not use the test synchronization for the ili test, if compiling for the pinetime. If used, the pinetime waits on stdio_rtt for 's' indefinitely.

Mind giving #13105 a spin on your PineTime with `CFLAGS += -DILI9341_PARAM_INVERTED=1 -DILI9341_PARAM_RGB=1`? :grin:

**Comment by kaspar030** [comment]
> Mind giving #13105 a spin on your PineTime with `CFLAGS += -DILI9341_PARAM_INVERTED=1 -DILI9341_PARAM_RGB=1`? grin

will do. the poor thing is going through a compile_and_test run ATM...

File: boards/pinetime/include/periph_conf.h (hunk 4 above)
Comment (kaspar030) [reply boards/pinetime/include/periph_conf.h]: folo :)

**Comment by kaspar030** [comment]
@bergzand this should be good to go?

File: boards/pinetime/Makefile.include [hunk 13]
```diff
@@ -0,0 +1,7 @@
+# for this board, we are using Segger's RTT as default terminal interface
+USEMODULE += stdio_rtt
```
Comment (aabadie) [inline boards/pinetime/Makefile.include]: Please have a look at #12724. That would help make things a little cleaner.

File: boards/common/nrf52xxxdk/board.c [hunk 14]
```diff
@@ -23,9 +23,11 @@
 
 void board_init(void)
 {
+#ifdef LED_PORT
```
Comment (aabadie) [inline boards/common/nrf52xxxdk/board.c]: Unrelated ?

File: boards/pinetime/include/gpio_params.h [hunk 15]
```diff
@@ -0,0 +1,38 @@
+/*
```
Comment (aabadie) [inline boards/pinetime/include/gpio_params.h]: The file is empty (e.g no saul gpio), why not just remove it (for now) ?

Comment (kaspar030) [reply boards/pinetime/include/gpio_params.h]: removed

File: boards/common/nrf52xxxdk/board.c (hunk 14 above)
Comment (kaspar030) [reply boards/common/nrf52xxxdk/board.c]: yup, removed

File: boards/pinetime/Makefile.include (hunk 13 above)
Comment (kaspar030) [reply boards/pinetime/Makefile.include]: Is this necessary if there's no periph_uart feature, which stdio_uart and ethos depend on?

Comment (aabadie) [reply boards/pinetime/Makefile.include]: I think so.

This board is in the same case as the hamilton, so things related to stdio_rtt could be replaced with just:
```mk
# use JLink Segger RTT by default
RIOT_TERMINAL ?= jlink
include $(RIOTMAKE)/tools/serial.inc.mk
```

And add the following in `pinetime/Makefile.dep`:
```mk
# Use Segger's RTT by default for stdio on this board
DEFAULT_MODULE += stdio_rtt
```

Comment (kaspar030) [reply boards/pinetime/Makefile.include]: This looks weird: `include $(RIOTMAKE)/tools/serial.inc.mk`.
The board doesn't have uarts exposed...

Comment (kaspar030) [reply boards/pinetime/Makefile.include]: Ok, now that I tried I realized that in master, this doesn't work. ;(

**Comment by emmanuelsearch** [comment]
Tested `examples/nimble_gatt` and `tests/driver_ili9341`. Both work like a charm!

**Comment by kaspar030** [comment]
I've done a full compile_and_test run. Many tests seem to run fine, but stdio_rtt output is pretty broken. Those that fail because of lost output actually succeeded.

I've managed to get unittests to pass using [this patch](https://gist.github.com/kaspar030/9e88623533f5733884349fe653d0aa27/raw) (adding an early xtimer_init call and increasing stdio_rtt output buffer size).

```
$ BOARD=pinetime make term
/home/kaspar/src/riot/dist/tools/jlink/jlink.sh term-rtt
### Starting RTT terminal ###
Twisted not available, please install it if you want to use pyterm's JSON capabilities
2020-01-14 15:38:54,102 # Host name for TCP connection is missing, defaulting to "localhost"
2020-01-14 15:38:54,102 # Connect to localhost:19021
Welcome to pyterm!
Type '/exit' to exit.
2020-01-14 15:38:55,108 # SEGGER J-Link V6.56b - Real time terminal output
2020-01-14 15:38:55,108 # SEGGER J-Link EDU V10.1, SN=260107551
2020-01-14 15:38:55,109 # Process: JLinkExe
2020-01-14 15:38:55,109 # main(): This is RIOT! (Version: 2020.01-devel-1724-ge7d34-add_pinetime_support)
2020-01-14 15:38:55,109 # Help: Press s to start test, r to print it is ready
s
2020-01-14 15:38:55,811 # START
2020-01-14 15:39:05,404 # ...............................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................................
2020-01-14 15:39:05,404 # OK (959 tests)
```

Otherwise the tests look fine. The stdio_rtt issues are unrelated to this board.

*ping* @bergzand, do you see more issues?

@aabadie, regarding the rtt configuration, IMO #12724 is independent of this PR, right?

@fjmolinas please keep an eye on this PR, would be awesome to have the support before fosdem!

**Comment by bergzand** [comment]
> I've managed to get unittests to pass using [this patch](https://gist.github.com/kaspar030/9e88623533f5733884349fe653d0aa27/raw) (adding an early xtimer_init call and increasing stdio_rtt output buffer size).

Have you tried compiling `stdio_rtt` with `STDIO_RTT_ENABLE_BLOCKING_STDOUT` enabled? Or is that unrelated here?

> ping @bergzand, do you see more issues?

I think you've covered everything from my side. Feel free to squash if the others agree.

**Comment by kaspar030** [comment]
> Have you tried compiling `stdio_rtt` with `STDIO_RTT_ENABLE_BLOCKING_STDOUT` enabled? Or is that unrelated here?

Yes... It didn't seem to help, there were still some output issues. I think we need to take a close look at stdio_rtt at some point.

**Comment by aabadie** [comment]
If you base your work on #12724 and disable the stdio_rtt module when `auto_init` is disabled, it will provide a simple workaround to the failing tests.
I managed to do a similar thing in #12304: https://github.com/RIOT-OS/RIOT/pull/12304/files#diff-cc423206e4b071a735d298217a657675R1029

**Comment by kaspar030** [comment]
> If you base your work on #12724 and disable the stdio_rtt module when `auto_init` is disabled, it will provide a simple workaround to the failing tests.

Talked offline, @aabadie is ok with not applying his suggestion.
IMO, the stdio_rtt issues are unrelated to the pinetime (there are boards in master that suffer the same issues, e.g., the ruuvitag). Also, #12724 can be seen as bugfix to the testing situation, so it is not bound by the feature freeze tomorrow. :wink: So I'll squash this now.

**Comment by bergzand** [comment]
@kaspar030 Murdock has some issues with missing docs for the ILI9341 defines. Feel free to amend directly.

**Comment by kaspar030** [comment]
> @kaspar030 Murdock has some issues with missing docs for the ILI9341 defines. Feel free to amend directly.

Yup, done. Murdock is happy, this needs an ACK!

**Comment by bergzand** [comment]
Al green, Go!

**Comment by emmanuelsearch** [comment]
Nice!

**Comment by kaspar030** [comment]
Thanks everyone!
